### PR TITLE
Add alerts to the sample Memcached operator

### PR DESCRIPTION
**Pull request by assafad**
Signed-off-by: assaf-admi <aadmi@redhat.com>

**Description of the change:**
This PR adds alerts and recording rule to the monitoring sample Memcached operator, including unit tests and runbooks for the new alerts.
Added alerts:
```
MemcachedDeploymentSizeUndesired
MemcachedOperatorDown
```
Added recording rule:
```
memcached_operator_up_total
```
In addition, this PR is addressing https://github.com/operator-framework/operator-sdk/issues/6140 as agreed.

**Motivation for the change:**
We would like to implement observability concepts for the Memcahced operator, in order to have the operator in [Deep Insights](https://sdk.operatorframework.io/docs/overview/operator-capabilities/#level-4---deep-insights) capability level. Then, we may use it as a reference in an observability best-practices tutorial in Operator-SDK. 

**Checklist**

If the pull request includes user-facing changes, extra documentation is required:
- [ ] Add a new changelog fragment in `changelog/fragments` (see [`changelog/fragments/00-template.yaml`](https://github.com/operator-framework/operator-sdk/tree/master/changelog/fragments/00-template.yaml))
- [ ] Add or update relevant sections of the docs website in [`website/content/en/docs`](https://github.com/operator-framework/operator-sdk/tree/master/website/content/en/docs)

### CODE REVIEW

**Comment by assafad** [comment]
Hi @machadovilaca, @sradco, this is an initial PR for the sample Memcached operator alerts, in order to hear your thoughts on the alerts, recording rule and unit tests that were added. It's probably missing a few steps before merging (e.g. not generated with `make generate` yet), and as @varshaprasad96 suggested, it may take place only in a separate monitoring sample operator.

**Comment by machadovilaca** [comment]
I like it, I think it looks good so far

File: testdata/go/v4-alpha/memcached-operator/monitoring/prom-rule-ci/prom-rules-tests.yaml
Comment (sradco) [inline testdata/go/v4-alpha/memcached-operator/monitoring/prom-rule-ci/prom-rules-tests.yaml:8]: @assafad Hi, Please look at this example https://github.com/kubevirt/kubevirt/blob/11316e05586b95dd3b3b4898f19cb2a743031e2c/hack/prom-rule-ci/prom-rules-tests.yaml#L71.

You can add additional values and test the result in different eval_time and merge some of the tests together.

File: testdata/go/v4-alpha/memcached-operator/monitoring/prom-rule-ci/prom-rules-tests.yaml
Comment (sradco) [inline testdata/go/v4-alpha/memcached-operator/monitoring/prom-rule-ci/prom-rules-tests.yaml:60]: @assafad Please remove empty lines

File: testdata/go/v4-alpha/memcached-operator/monitoring/alerts.go
Comment (sradco) [inline testdata/go/v4-alpha/memcached-operator/monitoring/alerts.go:56]: @assafad I think we should add example runbooks to the alerts.

File: testdata/go/v4-alpha/memcached-operator/go.sum
Comment (sradco) [inline testdata/go/v4-alpha/memcached-operator/go.sum:202]: @assafad I think we should remove these changes and only add the lines related to
```
github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring v0.60.1
```

File: testdata/go/v4-alpha/memcached-operator/go.sum
Comment (sradco) [inline testdata/go/v4-alpha/memcached-operator/go.sum:442]: @assafad same comment as above

File: testdata/go/v4-alpha/memcached-operator/go.mod
Comment (sradco) [inline testdata/go/v4-alpha/memcached-operator/go.mod:8]: @assafad I think we should only add this change and remove changes that are not related to it, unless they are dependencies of 
```
github.com/prometheus-operator/prometheus-operator/pkg/apis/monitoring v0.60.1
```

File: testdata/go/v4-alpha/memcached-operator/go.sum
Comment (assafad) [reply testdata/go/v4-alpha/memcached-operator/go.sum:202]: I believe these are all dependencies. I got an error about them while `docker build & push`

**Comment by assafad** [comment]
Hi @sradco, @machadovilaca, I updated the PR with a few changes:
- Added alerts runbooks
- Modified [prom-rules-tests.yaml](https://github.com/operator-framework/operator-sdk/pull/6135/files#diff-7f2eaf6ca58d6145dfbadcea55f3565d48538734770ca81ea2fce9c719d44cde)
- Since we now have a sample operator with default monitoring, I added to it some default RBAC in order to give Prometheus access to the operator's alerts. It assumes using kube-prometheus, as recommended in [exporting-metrics-for-prometheus](https://book.kubebuilder.io/reference/metrics.html#exporting-metrics-for-prometheus). The alerts tutorial will mention that, and will explain how to change it.
- Modified [memcached_controller.go](https://github.com/operator-framework/operator-sdk/pull/6135/files#diff-ecc9df479abf7d65e51c0ffdb46ec14c35055978bd85cc52c8f4b6f473207b52) to create PrometheusRule only when it doesn't exist yet.

Please have another look. One you will approve the changes, I will work to have it all scaffold.

**Comment by assafad** [comment]
Hi @varshaprasad96, @everettraven, could you please have a look?
This PR doesn't yet include changes in the samples generator, but only the suggested changes for the monitoring sample operator. Once the changes will be approved, I will work to have it scaffold.

File: testdata/go/v4-alpha/memcached-operator/monitoring/alerts.go
Comment (assafad) [reply testdata/go/v4-alpha/memcached-operator/monitoring/alerts.go]: Done

File: testdata/go/v4-alpha/memcached-operator/monitoring/prom-rule-ci/prom-rules-tests.yaml
Comment (assafad) [reply testdata/go/v4-alpha/memcached-operator/monitoring/prom-rule-ci/prom-rules-tests.yaml]: Done

File: testdata/go/v4-alpha/memcached-operator/monitoring/prom-rule-ci/prom-rules-tests.yaml
Comment (assafad) [reply testdata/go/v4-alpha/memcached-operator/monitoring/prom-rule-ci/prom-rules-tests.yaml]: Done

File: testdata/go/v4-alpha/monitoring/memcached-operator/test/e2e/e2e_test.go
Comment (everettraven) [inline testdata/go/v4-alpha/monitoring/memcached-operator/test/e2e/e2e_test.go:245]: Since the tests use `gomega`,  I *think* this could be simplified to one line with:
```go
Expect(1.0).Should(BeNumerically(">=", 1.0))
```
For more info on the `BeNumerically` function you can take a look at the [`gomega` go docs entry](https://pkg.go.dev/github.com/onsi/gomega#BeNumerically)

**Review by everettraven — APPROVED** [review]
@assafad Thanks for the contribution! Overall it looks good to me.

I have one tiny non-blocking nit:

File: testdata/go/v4-alpha/monitoring/memcached-operator/test/e2e/e2e_test.go
Comment (assafad) [reply testdata/go/v4-alpha/monitoring/memcached-operator/test/e2e/e2e_test.go]: Done

File: testdata/go/v4-alpha/monitoring/memcached-operator/controllers/memcached_controller.go
Comment (machadovilaca) [inline testdata/go/v4-alpha/monitoring/memcached-operator/controllers/memcached_controller.go:91]: not only you should check if the rule exists, but also if it was externally changed, and if so revert it to the initial configuration

File: testdata/go/v4-alpha/monitoring/memcached-operator/docs/runbooks/MemcachedDeploymentSizeUndesired.md
Comment (machadovilaca) [inline testdata/go/v4-alpha/monitoring/memcached-operator/docs/runbooks/MemcachedDeploymentSizeUndesired.md:4]: I think here it makes more sense to explain why the alert was triggered. More like "MemcachedDeploymentSizeUndesired is triggered when the number of available replicas of the Memcached deployment doesn't match the requested configuration"

File: testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go
Comment (machadovilaca) [inline testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go:15]: Will this be hardcoded in operator-sdk generator or are you thinking about configuring it somehow?

**Comment by machadovilaca** [comment]
would also like to see a test to verify all alerts have a runbook url

File: testdata/go/v4-alpha/monitoring/memcached-operator/docs/runbooks/MemcachedDeploymentSizeUndesired.md
Comment (assafad) [reply testdata/go/v4-alpha/monitoring/memcached-operator/docs/runbooks/MemcachedDeploymentSizeUndesired.md]: Thanks, done

File: testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go
Comment (assafad) [reply testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go:15]: @machadovilaca not sure I understood exactly what you are referring. Please explain more if possible.

File: testdata/go/v4-alpha/monitoring/memcached-operator/controllers/memcached_controller.go
Comment (assafad) [reply testdata/go/v4-alpha/monitoring/memcached-operator/controllers/memcached_controller.go:92]: I added it, please let me know the change makes sense.

File: testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go
Comment (machadovilaca) [reply testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go:15]: every new operator created will have a different runbookURLBasePath, right? My question was if you intended to leave this as `https://github.com/operator-framework/operator-sdk/tree...` by default and let the user change that later, or if you create it on operator code generation time

File: testdata/go/v4-alpha/monitoring/memcached-operator/controllers/memcached_controller.go
Comment (machadovilaca) [reply testdata/go/v4-alpha/monitoring/memcached-operator/controllers/memcached_controller.go:92]: looks good

File: testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go
Comment (assafad) [reply testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go:15]: Not sure what it has to change. The samples are stable on Github, no?
This code is only for the monitoring operator sample which will always have `memcached-operator` name, and its runbook base URL will always be `https://github.com/operator-framework/operator-sdk/tree/master/testdata/go/v4-alpha/monitoring/memcached-operator/runbooks/`. 
But please correct me if I'm wrong.

Comment (assafad) [reply testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go:15]: > every new operator created will have a different runbookURLBasePath

Why is that?

Comment (machadovilaca) [reply testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go:15]: in testdata yes, it won't change, my question is more for when you do the generator.

Comment (assafad) [reply testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go:15]: the generator is also for the samples, so don't see why this URL should be generated differently than other parts.
Basically, I planned to generate it hardcoded

Comment (machadovilaca) [reply testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go:15]: that generator is what operator-sdk uses to create new operator projects. What will those operators have in `runbookURLBasePath`?

Comment (assafad) [reply testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go:15]: AFAIK, the generator we're working on is only for the samples on testdata, which are not scafold by the default  operator-sdk scaffolding, no? 
https://github.com/operator-framework/operator-sdk/blob/master/hack/generate/samples/generate_testdata.go#L64

Comment (machadovilaca) [reply testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go:15]: mhh, maybe, but then users won't have the monitoring stuff if they generate a new project, right?

Comment (everettraven) [reply testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go:15]: @assafad @machadovilaca So my understanding on this PR is specifically to add the alerts and corresponding code *only* on the testdata. We have a generator specifically for generating our testdata that @assafad is updating and not the actual generation logic used in the `operator-sdk` cli tool that users would use to scaffold their projects.

That being said, I think @machadovilaca points do raise the question - If the plan is to ultimately create a plugin that will do most of the heavy lifting by scaffolding most of this logic, should we really be updating the testdata prior to creating that plugin? It feels to me like it may be better to go ahead and start writing the plugin and tackle updating the testdata once the plugin is written so you aren't writing testdata generation logic just to delete it later when the plugin automatically scaffolds *most* of that logic for you.

Comment (machadovilaca) [reply testdata/go/v4-alpha/monitoring/memcached-operator/monitoring/alerts.go:15]: we actually just met to discuss this, and we agree with you, it does not make sense to update the testdata generator, but to first create a plugin in kubebuilder instead

**Comment by sradco** [comment]
> > would also like to see a test to verify all alerts have a runbook url
> 
> @machadovilaca I plan to have another PR with e2e testing, which includes alerts test and rules tests (verifying each alert has `summary`, `severity`, accessible `runbook_url`).

runbook_url is optional , so we will not require it in the alert fields test.

**Comment by varshaprasad96** [comment]
@assafad The contents added to the sample project looks good. What I am concerned about is that these are in the testdata/ directly, instead they need to be in the generator or else re-generating the testdata will wipe these changes. Another addition which will be really useful is to add documentation in the SDK website (https://sdk.operatorframework.io), about this sample monitoring project, the additions made and benefits for anyone who is looking to add a prometheus monitoring stack.

**Comment by assafad** [comment]
> @assafad The contents added to the sample project looks good. What I am concerned about is that these are in the testdata/ directly, instead they need to be in the generator or else re-generating the testdata will wipe these changes. Another addition which will be really useful is to add documentation in the SDK website (https://sdk.operatorframework.io), about this sample monitoring project, the additions made and benefits for anyone who is looking to add a prometheus monitoring stack.

@varshaprasad96 thanks. I'm now working on generating the changes which are proposed. Regarding your suggestion for documentation, we have an [issue](https://issues.redhat.com/browse/CNV-21038) for that, and will address it in a different PR.

File: testdata/go/v4-alpha/monitoring/memcached-operator/go.mod
Comment (varshaprasad96) [inline testdata/go/v4-alpha/monitoring/memcached-operator/go.mod:10]: Can we avoid bumping k8s version in this PR? This is because the k8s bumps (minor or major) are usually tracked separately so as changes happening due to them.

File: testdata/go/v4-alpha/monitoring/memcached-operator/go.mod
Comment (varshaprasad96) [inline testdata/go/v4-alpha/monitoring/memcached-operator/go.mod:13]: Same with controller-runtime

File: testdata/go/v3/monitoring/memcached-operator/monitoring/prom-rule-ci/prom-rule-tests.yaml
Comment (assafad) [reply testdata/go/v3/monitoring/memcached-operator/monitoring/prom-rule-ci/prom-rule-tests.yaml]: Added more explanations and a link to Prometheus official documentation

File: testdata/go/v4-alpha/monitoring/memcached-operator/go.mod
Comment (assafad) [reply testdata/go/v4-alpha/monitoring/memcached-operator/go.mod:10]: I'm not sure. Maybe it's running `go mod tidy` after generating? How can I avoid these bumps?

File: hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go
Comment (avlitman) [inline hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go:768]:
```suggestion
	makefilePath := filepath.Join(mh.ctx.Dir, "Makefile")
```

File: hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go
Comment (avlitman) [inline hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go:771]:
```suggestion
	err := kbutil.InsertCode(makefilePath,
```

**Review by avlitman — CHANGES_REQUESTED** [review]

File: hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go
Comment (assafad) [reply hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go]: thanks, done

File: hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go
Comment (assafad) [reply hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go]: thanks, done

File: hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go
Comment (avlitman) [inline hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go:532]:
```suggestion
	metricsPath := filepath.Join(mh.ctx.Dir, "monitoring/metrics.go")
	_, err := os.Create(metricsPath)
	pkg.CheckError("creating metrics file", err)

```

Comment (avlitman) [reply hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go:532]: You use `filepath.Join(mh.ctx.Dir, "monitoring/metrics.go")` twice, no need (Joao comment this on my pr as well)

File: hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go
Comment (avlitman) [inline hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go:552]:
```suggestion
	alertsPath := filepath.Join(mh.ctx.Dir, "monitoring/alerts.go")
	_, err := os.Create(alertsPath)
	pkg.CheckError("creating alerts file", err)

```

Comment (assafad) [reply hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go]: Done for all created files, thanks!

File: hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go
Comment (assafad) [reply hack/generate/samples/internal/go/memcached-with-webhooks/memcached_with_webhooks.go]: Done for all created files, thanks!

File: testdata/go/v3/monitoring/memcached-operator/controllers/memcached_controller.go
Comment (varshaprasad96) [inline testdata/go/v3/monitoring/memcached-operator/controllers/memcached_controller.go:48]: can we get this from `req.NamespacedName`? This contains the name and namespace of the object that would be reconciled. Should the `monitoringv1.PrometheusRule` be in the same location?

Comment (assafad) [reply testdata/go/v3/monitoring/memcached-operator/controllers/memcached_controller.go:48]: I wasn't able to get this from `req.NamespacedName` which AFAIK refers to `memcached` here. 
Yes, we create the `monitoringv1.PrometheusRule` in the same namespace.

**Comment by varshaprasad96** [comment]
The PR looks good to me! Thanks for the contribution @assafad!
Waiting for another set of eyes to review before merging. cc: @everettraven @camilamacedo86 @Kavinjsir @theishshah

**Review by jberkhahn — APPROVED** [review]

**Pull request merged**
